### PR TITLE
Allow direct compilation of a Policy into a valid Descriptor

### PR DESCRIPTION
**Pull request by SarcasticNastik**
This PR works on top of  #291 and #342. It introduces a new function `compile_to_descriptor` which requires a `DescriptorCtx` (suggestions for a better enum name are welcome!) allowing directly compilation of a policy into the specified descriptor type.

### CODE REVIEW

File: src/policy/concrete.rs [hunk 1]
```diff
@@ -108,6 +108,20 @@ pub enum PolicyError {
...
+pub enum DescriptorCtx<Pk> {
+    /// Bare
+    Bare,
+    /// Sh
+    Sh,
+    /// Wsh
+    Wsh,
+    /// ShWsh
+    ShWsh,
+    /// Tr
```
Comment (sanket1729) [inline src/policy/concrete.rs:121]: This should mention that Option<Pk> is internal key

**Comment by sanket1729** [comment]
I will review this after we get the first PR in.

**Comment by sanket1729** [comment]
This can be rebased again after v2 one. Let's try to merge this before thresh enumeration

**Comment by sanket1729** [comment]
Merged v2 one, this can be rebased on top of master now

**Review by apoelstra — APPROVED** [review]
ACK f464e333823d31a75ad899bea5287cd6f6e597ba

File: src/policy/concrete.rs [hunk 2]
```diff
@@ -100,6 +100,20 @@ pub enum PolicyError {
...
+pub enum DescriptorCtx<Pk> {
+    /// [Bare][`Descriptor::Bare`]
+    Bare,
+    /// [Sh][`Descriptor::Sh`]
+    Sh,
+    /// [Wsh][`Descriptor::Wsh`]
+    Wsh,
+    /// Sh-wrapped [Wsh][`Descriptor::Wsh`]
+    ShWsh,
+    /// [Tr][`Descriptor::Tr`], where the Option<Pk> corresponds to the internal_key.
```
Comment (sanket1729) [inline src/policy/concrete.rs:113]: the documentation here should highlight that his is used as the internal key only when there is no inferred from Policy

File: src/policy/concrete.rs [hunk 3]
```diff
@@ -289,6 +303,28 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
         }
     }
 
+    /// Compile the [`Policy`] into desc_ctx [`Descriptor`]
+    #[cfg(feature = "compiler")]
+    pub fn compile_to_descriptor<Ctx: ScriptContext>(
```
Comment (sanket1729) [inline src/policy/concrete.rs:308]: Same doccoment about `internal_key` here.

**Review by sanket1729 — COMMENTED** [review]
Left two small documentation comments

**Comment by SarcasticNastik** [comment]
@sanket1729 Done!

**Review by apoelstra — APPROVED** [review]
ACK 169d8498567e33b5ccf9ff37788884538914c72d

**Pull request merged**